homeassistant/components/plugwise/climate.py

```diff
@@ -0,0 +1,250 @@
+from homeassistant.components.climate import PLATFORM_SCHEMA, ClimateDevice
+from homeassistant.components.climate.const import (
+    CURRENT_HVAC_HEAT,
+    CURRENT_HVAC_IDLE,
+    HVAC_MODE_AUTO,
+    HVAC_MODE_OFF,
+    SUPPORT_PRESET_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+)
+from homeassistant.const import (
+    ATTR_TEMPERATURE,
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_USERNAME,
+    TEMP_CELSIUS,
+)
+from homeassistant.exceptions import PlatformNotReady
+
+SUPPORT_FLAGS = SUPPORT_TARGET_TEMPERATURE | SUPPORT_PRESET_MODE
+
+_LOGGER = logging.getLogger(__name__)
+
+# Configuration directives
+CONF_MIN_TEMP = "min_temp"
+CONF_MAX_TEMP = "max_temp"
+
+# Default directives
+DEFAULT_NAME = "Plugwise Thermostat"
+DEFAULT_USERNAME = "smile"
+DEFAULT_TIMEOUT = 10
+DEFAULT_PORT = 80
+DEFAULT_ICON = "mdi:thermometer"
+DEFAULT_MIN_TEMP = 4
+DEFAULT_MAX_TEMP = 30
+
+# HVAC modes
+ATTR_HVAC_MODES = [HVAC_MODE_AUTO, HVAC_MODE_OFF]
+
+# Read platform configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_USERNAME, default=DEFAULT_USERNAME): cv.string,
+        vol.Optional(CONF_MIN_TEMP, default=DEFAULT_MIN_TEMP): cv.positive_int,
+        vol.Optional(CONF_MAX_TEMP, default=DEFAULT_MAX_TEMP): cv.positive_int,
+    }
+)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Add the Plugwise (Anna) Thermostate."""
+    add_devices(
+        [
+            ThermostatDevice(
+                config.get(CONF_NAME),
+                config.get(CONF_USERNAME),
+                config.get(CONF_PASSWORD),
+                config.get(CONF_HOST),
+                config.get(CONF_PORT),
+                config.get(CONF_MIN_TEMP),
+                config.get(CONF_MAX_TEMP),
+            )
+        ]
+    )
+
+
+class ThermostatDevice(ClimateDevice):
+    """Representation of an Plugwise thermostat."""
+
+    def __init__(self, name, username, password, host, port, min_temp, max_temp):
+        """Set up the Plugwise API."""
+        _LOGGER.debug("Init called")
+        self._name = name
+        self._username = username
+        self._password = password
+        self._host = host
+        self._port = port
+        self._temperature = None
+        self._current_temperature = None
+        self._outdoor_temperature = None
+        self._state = None
+        self._active_schema = None
+        self._previous_schema = None
+        self._preset_mode = None
+        self._preset_modes = []
+        self._min_temp = min_temp
+        self._max_temp = max_temp
+        self._hvac_modes = ATTR_HVAC_MODES
+        self._hvac_mode = None
+        self._available_schemas = []
+
+        _LOGGER.debug("Initializing API")
+        import haanna
```

Comment:
3rd party imports can be moved to the top of the module.

Reply:
There was a reason why it was moved down (in `custom_component` time) don't recall why. Fixed, tested (negative and positive) it accordingly.

Reply:
Before the time of `manifest.json` we had to import 3rd party integration specific requirements inside functions since those were installed on demand after loading the modules of the integration. Now we install the requirements on demand but before loading the modules.

Reply:
Good to know, I stepped into this (hopefully) former `custom_component` right about when `manifest.json` was being used.